tools/clang/lib/CodeGen/CGHLSLMS.cpp

```diff
@@ -5069,6 +5069,57 @@ static void CreateWriteEnabledStaticGlobals(llvm::Module *M,
   }
 }
 
+// Translate RayQuery constructor.  From:
+//  %call = call %"RayQuery<flags>" @<constructor>(%"RayQuery<flags>" %ptr)
+// To:
+//  i32 %handle = AllocateRayQuery(i32 <IntrinsicOp::IOP_AllocateRayQuery>, i32 %flags)
+//  %gep = GEP %"RayQuery<flags>" %ptr, 0, 0
+//  store i32* %gep, i32 %handle
+//  ; and replace uses of %call with %ptr
+void TranslateRayQueryConstructor(llvm::Module &M) {
+  SmallVector<Function*, 4> Constructors;
+  for (auto &F : M.functions()) {
+    // Match templated RayQuery constructor instantiation by prefix and signature.
+    // It should be impossible to achieve the same signature from HLSL.
+    if (!F.getName().startswith("\01??0?$RayQuery@$"))
+      continue;
+    llvm::Type *Ty = F.getReturnType();
+    if (!Ty->isPointerTy() || !dxilutil::IsHLSLRayQueryType(Ty->getPointerElementType()))
+      continue;
```

Comment:
This is extra caution for if something changes, my mangled prefix assumption was incorrect, another constructor is added that should be handled elsewhere, or something unanticipated.  It will fall through, fail validation, and be visibly something more obvious that was unhandled.  Other approaches exist... we could assert, then crash if something accessed is not expected (on release build), but that might let some unexpected runtime behavior through on release, such as additional args (new constructor) getting ignored or something like that.  This seemed to be a decent way to be sure I'm only translating the expected thing here.